src/binary-writer.cc

```diff
@@ -468,6 +527,90 @@ static uint8_t log2_u32(uint32_t x) {
   return result;
 }
 
+template <ExprType T>
+const Expr* BinaryWriter::BlockState<T>::advance(class BinaryWriter& writer,
+                                                 const Func* func) {
+  if (this->current_expr != this->expr->block.exprs.end()) {
+    return this->get_current_and_advance();
+  }
+  WriteOpcode(writer.stream_, Opcode::End);
+  return nullptr;
+}
+
+const Expr* BinaryWriter::IfState::advance(class BinaryWriter& writer,
+                                           const Func* func) {
+  if (!in_else) {
+    if (current_expr != expr->true_.exprs.end()) {
+      return get_current_and_advance();
+    }
+    if (!expr->false_.empty()) {
+      WriteOpcode(writer.stream_, Opcode::Else);
+    }
+    current_expr = expr->false_.begin();
+    in_else = true;
+  }
+  if (current_expr != expr->false_.end()) {
+    return get_current_and_advance();
+  }
+  WriteOpcode(writer.stream_, Opcode::End);
+  return nullptr;
+}
+
+const Expr* BinaryWriter::TryState::advance(class BinaryWriter& writer,
+                                            const Func* func) {
+  if (!in_catch) {
+    if (current_expr != expr->block.exprs.end()) {
+      return get_current_and_advance();
+    }
+    switch (expr->kind) {
+      case TryKind::Catch:
+        in_catch = true;
+        break;
+      case TryKind::Delegate:
+        WriteOpcode(writer.stream_, Opcode::Delegate);
+        WriteU32Leb128(writer.stream_,
+                       writer.GetLabelVarDepth(&expr->delegate_target),
+                       "delegate depth");
+        return nullptr;
+      case TryKind::Plain:
+        WriteOpcode(writer.stream_, Opcode::End);
+        return nullptr;
+    }
+  }
+  // deliberately no else
+  if (in_catch) {
+    if (current_catch < expr->catches.size()) {
+      auto& catch_block = expr->catches[current_catch];
+      if (current_expr != catch_block.exprs.end()) {
+        if (current_expr == catch_block.exprs.begin()) {
+          if (catch_block.IsCatchAll()) {
+            WriteOpcode(writer.stream_, Opcode::CatchAll);
+          } else {
+            WriteOpcode(writer.stream_, Opcode::Catch);
+            WriteU32Leb128(writer.stream_,
+                           writer.GetTagVarDepth(&catch_block.var),
+                           "catch tag");
+          }
+        }
+        return get_current_and_advance();
+      }
+      do {
+        current_catch++;
+      } while (current_catch < expr->catches.size() &&
+               expr->catches[current_catch].exprs.empty());
```

Comment:
Don't skip empty `catch`, it means wasm author wanted to catch exception and ignore it.